### PR TITLE
Creating Tensile Documentation 

### PR DESCRIPTION
**Pull request by SwRaw**
Updates style, grammar, and clarity of Tensile's ROCm docs, and migrates important files from the Wiki.

### CODE REVIEW

File: docs/src/getting-started/installation.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+********************************************************************
+
+You can obtain Tensile as part of ROCm installation or choose to install Tensile from source. This document discusses both the methods to obtain Tensile.
+
+Install ROCm
+============
+
+1. Install ROCm for your platform. For installation instructions, refer to `Linux <https://rocm.docs.amd.com/projects/install-on-linux/en/latest/tutorial/quick-start.html>`_ or Windows `Windows <https://rocm.docs.amd.com/projects/install-on-windows/en/latest/index.html>` installation guide.
+
+   After the installation is complete, you can find the binaries and libraries in ``/opt/rocm``. Installing ROCm provides compilers such as ``amdclang++``, and other useful tools including ``rocminfo`` and ``rocprofv2``.
```
Comment (SwRaw) [inline docs/src/getting-started/installation.rst:18]: @bstefanuk Is this information related to Tensile installation in any way?

File: docs/src/getting-started/installation.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+   .. tip::
+
+      If using Bash, we recommend you to set ``PATH=/opt/rocm/bin/:$PATH`` in your ``~/.bashrc`` and refresh your shell, using ``source ~/.bashrc``.
+      Alternatively, export the path for your current shell session only, using ``export PATH=/opt/rocm/bin/:$PATH``.
+
+Install OS dependencies
+=======================
+
+Here are the steps to install dependencies for Ubuntu distributions. For other distributions, use the appropriate package manager.
```
Comment (SwRaw) [inline docs/src/getting-started/installation.rst:28]: @bstefanuk For other distributions, where can the users find the appropriate package manager? Can we provide ROCm package manager link here?

File: docs/src/getting-started/installation.rst [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+   apt-get install libtinfo-dev      # If using the YAML backend
+
+3. Install build tools. For additional installation methods for the latest versions of CMake, see the `CMake installation <https://cliutils.gitlab.io/modern-cmake/chapters/intro/installing.html>`_ page.
+
+   .. code-block::
+
+      apt-get install build-essential cmake
+
+4. Verify the versions of installed tools against the following table:
```
Comment (SwRaw) [inline docs/src/getting-started/installation.rst:56]: @bstefanuk What does the user need to verify? Should the installed tools match the versions mentioned in the table? The action item for user is not clear here.

File: docs/src/getting-started/installation.rst [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+Option 2: Install with git
+----------------------------
+
+.. code-block:: bash
+
+  git clone git@github.com:ROCm/Tensile.git && cd Tensile
+  pip3 install .
+
+You can now run Tensile's Python applications—see :ref:`quick-start`.
```
Comment (SwRaw) [inline docs/src/getting-started/installation.rst:97]: @bstefanuk Is quick start the only way to run Tensile's python applications? Normally a quick start is a shortcut to achieve something while a longer method is also available. I am trying to understand if quick start is actually a quick start? If its the only "start" then let's not call it quick start or otherwise we need to mention the other method to run applications as well.

File: docs/src/how-to/programmers-guide.rst [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+=======================
+Development environment
+=======================
+
+ROCm is the base requirement for contributing to Tensile. See if ROCm is supported on your platform by verifying the `supported operating systems <https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/system-requirements.html#supported-operating-systems>`_ list.
+Then, follow the steps given in the :ref:`installation` guide.
+
+=====================
+Developing in Docker
```
Comment (SwRaw) [inline docs/src/how-to/programmers-guide.rst:24]: @bstefanuk Is this section under Development environment? 
Also, is ROCm installation required if using docker?

File: docs/src/how-to/programmers-guide.rst [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+  Running the preceding command generates an executable ``TensileTests``, which can be further used to run the tests.
+
+- Manually: To build and run the tests manually, see the commands in `tasks.py <https://github.com/ROCm/Tensile/blob/develop/tasks.py>`_.
+  For advanced usage like filtering or repeating test cases, see the `gtest documentation <https://github.com/google/googletest/blob/main/docs/advanced.md>`_.
+
+===============
+Static analysis
+===============
+
```
Comment (SwRaw) [inline docs/src/how-to/programmers-guide.rst:131]: @bstefanuk We need some information about static analysis here. Also some sentences to explain the Python and C++ classification.

File: docs/src/how-to/programmers-guide.rst [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+.. code-block::
+
+   clang-format -i style=file <files>
+
+Styling rules are configured in `.clang-format <https://github.com/ROCm/Tensile/blob/develop/.clang-format>`_.
+
+=========
+Profiling
+=========
+
```
Comment (SwRaw) [inline docs/src/how-to/programmers-guide.rst:174]: @bstefanuk Need information about Profiling. Also, some sentences to understand how Python is related?

File: docs/src/how-to/programmers-guide.rst [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+To enable profiling, use the ``@profile`` decorator, which must be imported from the ``Tensile.Utilities.Profile`` module. Under the hood, the decorator wraps the function in a `cProfile <https://docs.python.org/3/library/profile.html#module-cProfile>`_ context and generates a ``.prof`` file inside the ``profiling-results-<date>`` directory.
+
+.. note::
+   Nested profiling is NOT supported due to the existing limitation with the profiling decorator. This implies that if `func1` calls `func2` in a loop, and both are marked for profiling, the resulting ``.prof`` file for `func1` will display incorrect results.
+
+========================
+Building documentation
+========================
+
+Tensile uses https://github.com/ROCm/rocm-docs-core as the documentation engine, which itself wraps Read the Docs and Sphinx.
```
Comment (SwRaw) [inline docs/src/how-to/programmers-guide.rst:188]: @bstefanuk Can we replace this para with just the link https://rocm.docs.amd.com/en/latest/contribute/contributing.html?
I am not sure if we need to talk about rocm-docs-core here as users don't have to do anything there.

File: CONTRIBUTING.rst [hunk 9]
```diff
@@ -1,35 +1,32 @@
...
-2. Clone your fork locally and set up your :ref:`development-environment`.
-3. Create a feature branch off of **develop** and make changes to the code.
-4. Issue ``tox run -m precommit`` and ensure that all checks pass.
-5. Commit you changes using the convention for :ref:`commit-messages`.
-6. If you are updating documentation, issue ``tox run -e docs`` and verify the styling and formatting is what you expect.
-7. Push the changes to your fork.
+   1. Create a fork of Tensile. Do not create feature branches directly in https://github.com/ROCm/Tensile.
+   2. Clone your fork locally and set up your :ref:`development-environment`.
+   3. Create your feature branch from **develop** and make changes to the code.
+   4. Issue ``tox run -m precommit`` and ensure that all checks pass.
```
Comment (SwRaw) [inline CONTRIBUTING.rst:22]: @bstefanuk We have provided some related information in programmer's guide as well, but I couldn't find this command there. Is there any overlap that can be avoided?

File: CONTRIBUTING.rst [hunk 10]
```diff
@@ -1,54 +1,51 @@
...
-**When opening a PR:**
+b. **Creating the PR:**
 
-1. Ensure that **your develop** branch is up-to-date with the **upstream develop** branch---this may require a rebase or a merge.
-2. Verify that your changes pass static analysis checks and all pre-checkin, host library, and unit tests by running ``tox run -m prepr``---then go get a coffee, this could take up to an hour.
-3. Create the PRs against the https://github.com/ROCm/Tensile **develop** branch.
-4. Fill in as many details as possible. Include a description, outcomes, notable changes, and environment information. This more information, the more likely the PR will be reviewed and merged in a timely manner.
-5. Title the PR in present imperative tense, e.g., "*Update* kernel parameters" not "Updates" nor "Updated".
+   1. Ensure that **your develop** branch is up-to-date with the **upstream develop** branch. This might require a rebase or a merge.
+   2. Verify that your changes pass static analysis checks and all pre-checkin, host library, and unit tests by running ``tox run -m prepr``.
```
Comment (SwRaw) [inline CONTRIBUTING.rst:34]: @bstefanuk We have provided some related information in programmer's guide as well, but I couldn't find this command there. Is there any overlap that can be avoided?

File: docs/src/getting-started/installation.rst [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+********************************************************************
+Installation
+********************************************************************
+
+You can obtain Tensile as part of ROCm installation or choose to install Tensile from source. This document discusses both the methods to obtain Tensile.
+
+Install ROCm
+============
+
+1. Install ROCm for your platform. For installation instructions, refer to `Linux <https://rocm.docs.amd.com/projects/install-on-linux/en/latest/tutorial/quick-start.html>`_ or Windows `Windows <https://rocm.docs.amd.com/projects/install-on-windows/en/latest/index.html>` installation guide.
```
Comment (bstefanuk) [inline docs/src/getting-started/installation.rst:16]:
```suggestion
1. Install ROCm for your platform. For installation instructions, refer to `Linux <https://rocm.docs.amd.com/projects/install-on-linux/en/latest/tutorial/quick-start.html>`_ or Windows `Windows <https://rocm.docs.amd.com/projects/install-on-windows/en/latest/index.html>`_ installation guide.
```

File: docs/src/getting-started/installation.rst (hunk 1 above)
Comment (bstefanuk) [reply docs/src/getting-started/installation.rst:18]: I would say we should move line 18 to the Programmer's Guide and update it to day
```suggestion
   After the installation is complete, ROCm compilers such as ``amdclang++`` can be found in ``/opt/rocm``, which is called implicitly throughout the various Tensile applications.
```

Comment (bstefanuk) [reply docs/src/getting-started/installation.rst:18]: For line 18, no, we can remove it. But please keep step number 1 for completeness.

File: docs/src/getting-started/installation.rst (hunk 2 above)
Comment (bstefanuk) [reply docs/src/getting-started/installation.rst:28]: This would be a significant addition that probably won't add much value. Typically those who work on non-Ubuntu OS's such as RHEL will be familiar with the package manager of their OS. The following page (linked in the docs in step 1) show the package managers for the different OSs used when installing ROCm. That could be used as a reference.

File: docs/src/getting-started/installation.rst (hunk 3 above)
Comment (bstefanuk) [reply docs/src/getting-started/installation.rst:56]: Please remove this line -- out-dated.

File: docs/src/getting-started/installation.rst (hunk 4 above)
Comment (bstefanuk) [reply docs/src/getting-started/installation.rst:97]: We have road mapped the addition of example docs describing how to run more advanced workflows, but that's still a few sprints away. Let's keep the quick start how it is for now, and I'll add the link here when the subsequent docs are added.

File: docs/src/how-to/programmers-guide.rst (hunk 5 above)
Comment (bstefanuk) [reply docs/src/how-to/programmers-guide.rst:24]: Yes it should be under Development  environment, typically it's best to setup a docker-based workflow for development so it's a good idea to keep it here.

ROCm should come installed in AMDs docker images.

File: docs/src/how-to/programmers-guide.rst (hunk 6 above)
Comment (bstefanuk) [reply docs/src/how-to/programmers-guide.rst:131]: This is WIP. For you, you can either remove it entirely or add a callout that this section is under development.

File: docs/src/how-to/programmers-guide.rst (hunk 7 above)
Comment (bstefanuk) [reply docs/src/how-to/programmers-guide.rst:174]: This section is the profiling section, Python is already added, Cpp will come later (same as above). You can leave this as is.

**Comment by bstefanuk** [comment]
@SwRaw changes are looking good. Once you apply the comments, please resolve the items and I'll do a final review.

File: docs/src/getting-started/installation.rst (hunk 1 above)
Comment (SwRaw) [reply docs/src/getting-started/installation.rst:18]: Done.

File: docs/src/getting-started/installation.rst (hunk 2 above)
Comment (SwRaw) [reply docs/src/getting-started/installation.rst:28]: Ok. Then, let's leave it like that.

File: docs/src/getting-started/installation.rst (hunk 3 above)
Comment (SwRaw) [reply docs/src/getting-started/installation.rst:56]: Removed point 4.

File: docs/src/getting-started/installation.rst (hunk 4 above)
Comment (SwRaw) [reply docs/src/getting-started/installation.rst:97]: I realized that a separate quick start file doesn't make sense since it is in continuation with installation and this quick start is not a quick install but a quick way of running benchmarks.
I have made the changes.

File: docs/src/how-to/programmers-guide.rst (hunk 5 above)
Comment (SwRaw) [reply docs/src/how-to/programmers-guide.rst:24]: Done.

File: docs/src/how-to/programmers-guide.rst (hunk 6 above)
Comment (SwRaw) [reply docs/src/how-to/programmers-guide.rst:131]: ok.

File: docs/src/how-to/programmers-guide.rst (hunk 7 above)
Comment (SwRaw) [reply docs/src/how-to/programmers-guide.rst:174]: ok

File: docs/src/how-to/programmers-guide.rst (hunk 8 above)
Comment (SwRaw) [reply docs/src/how-to/programmers-guide.rst:188]: Done.

File: CONTRIBUTING.rst (hunk 9 above)
Comment (SwRaw) [reply CONTRIBUTING.rst]: @bstefanuk Please respond to this as well.

File: CONTRIBUTING.rst (hunk 10 above)
Comment (SwRaw) [reply CONTRIBUTING.rst]: @bstefanuk Please respond to this as well.

File: CONTRIBUTING.rst (hunk 9 above)
Comment (bstefanuk) [reply CONTRIBUTING.rst]: We should leave this for completeness. Not much overlap and I think this is good to keep in the contributor guide.

File: CONTRIBUTING.rst (hunk 10 above)
Comment (bstefanuk) [reply CONTRIBUTING.rst]: Same as above, this is fine to leave in the contributor guide.

File: docs/src/cli-reference/tensile_create_library_cli.rst [hunk 12]
```diff
@@ -0,0 +1,127 @@
+.. meta::
+  :description: Tensile documentation and API reference
+  :keywords: Tensile, GEMM, Tensor, ROCm, API, Documentation
+
+.. _tensilecreatelibrary-cli-reference:
+
+====================
+TensileCreateLibrary
+====================
+
```
[the remark this answers is not in the document]
Comment (bstefanuk) [reply docs/src/cli-reference/tensile_create_library_cli.rst:10]: The CLI tool can use the following description:

> TensileCreateLibrary generates libraries and code object files for a set of supplied logic files.

File: docs/src/cli-reference/tensile_create_library_cli.rst [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+  * - Option
+    - Description
+
+  * - \-\-architecture=ARCHITECTURE
+    - Architectures to generate a library for. When specifying multiple options, use quoted and semicolon-delimited
+      architectures such as \-\-architecture='gfx908;gfx1012'.
+      Supported architectures include: all; gfx000; gfx803; gfx900; gfx900:xnack-; gfx906; gfx906:xnack+; gfx906:xnack-; gfx908; gfx908:xnack+;
+      gfx908:xnack-; gfx90a; gfx90a:xnack+; gfx90a:xnack-; gfx940; gfx940:xnack+; gfx940:xnack-; gfx941; gfx941:xnack+;
+      gfx941:xnack-; gfx942; gfx942:xnack+; gfx942:xnack-; gfx1010; gfx1011; gfx1012; gfx1030; gfx1031; gfx1032; gfx1034; gfx1035;
+      gfx1100; gfx1101; gfx1102.
```
Comment (SwRaw) [inline docs/src/cli-reference/tensile_create_library_cli.rst:59]: @bstefanuk Are the supported architectures the same as the supported architectures for ROCm? If yes, then we don't need to mention these.

Comment (bstefanuk) [reply docs/src/cli-reference/tensile_create_library_cli.rst:59]: No they're not, Tensile supports a sub-set of available cards, and it often takes some time for us to bring support for new architectures, so this is important information.

**Comment by bstefanuk** [comment]
> Sorry, I'm confused. Where is documentation hosted? It's not the same as the wiki on github?

@pghysels We're in the process of moving Tensile's Wiki into ROCm docs. As a part of this effort, we're working on restructuring some of the documents to match other library standards. The links you provided appear to be generated from the Wiki content, but I'm not familiar with who has published these, I suspect they're unofficial.

Until this migration is complete, the official AMD docs are hosted privately.

**Comment by SwRaw** [comment]
> Sorry, I'm confused. Where is documentation hosted? It's not the same as the wiki on github? I found these: https://alt-doc-doc-test.readthedocs.io/en/latest/ROCm_Tools/tensile.html https://new-documentation.readthedocs.io/en/latest/ROCm_Tools/tensile.html#contributing which are also different?

You can see the docs here: https://advanced-micro-devices-demo--1999.com.readthedocs.build/projects/Tensile/en/1999/src/index.html

File: docs/src/api-reference/tensile_create_library.rst [hunk 14]
```diff
@@ -1,24 +1,27 @@
...
+  :description: Tensile documentation and API reference
+  :keywords: Tensile, GEMM, Tensor, ROCm, API, Documentation
 
 .. _tensilecreatelibrary-api-reference:
 
-====================
-TensileCreateLibrary
-====================
+======================
+TensileCreate library
```
Comment (bstefanuk) [inline docs/src/api-reference/tensile_create_library.rst:8]:
```suggestion
TensileCreateLibrary
```

File: docs/src/api-reference/embedded_data.rst [hunk 15]
```diff
@@ -0,0 +1,11 @@
+.. meta::
+  :description: Tensile documentation and API reference
+  :keywords: Tensile, GEMM, Tensor, ROCm, API, Documentation
+
+.. _embeddeddata-api-reference:
+
+==============
+Embedded Data
```
Comment (bstefanuk) [inline docs/src/api-reference/embedded_data.rst:8]:
```suggestion
EmbeddedData
```

**Comment by bstefanuk** [comment]
@SwRaw A few things I noticed regarding the following picture of the side bar:
image

Can we move the **CLI reference** dropdown outside of the **API reference** please. Also, let's make **API reference** a drop-down too and nest them both under the **Reference** section.

**Comment by SwRaw** [comment]
@bstefanuk I am struggling to create the nested TOC structure as requested by you. Let me know if you have any idea on how to nest captions.

**Review by bstefanuk — APPROVED** [review]
Good updates. Thank you.
